src/nvim/tui/tui.c

```diff
@@ -2078,10 +2117,7 @@ static void augment_terminfo(TUIData *data, const char *term, long vte_version,
   data->unibi_ext.disable_extended_keys = unibi_find_ext_str(ut, "Dseks");
   if (data->unibi_ext.enable_extended_keys == -1) {
     if (!kitty && (vte_version == 0 || vte_version >= 5400)) {
-      data->unibi_ext.enable_extended_keys = (int)unibi_add_ext_str(ut, "ext.enable_extended_keys",
-                                                                    "\x1b[>4;2m");
-      data->unibi_ext.disable_extended_keys = (int)unibi_add_ext_str(ut, "ext.disable_extended_keys",
-                                                                     "\x1b[>4m");
+      data->input.supports_modify_other_keys = true;
     }
   }
```

Comment:
Shouldn't it be as follows?

```C  
  if (data->unibi_ext.enable_extended_keys == -1) {
    data->unibi_ext.enable_extended_keys = (int)unibi_add_ext_str(ut, "ext.enable_extended_keys",
                                                                  "\x1b[>4;2m");
  }
  if (data->unibi_ext.disable_extended_keys == -1) {
    data->unibi_ext.disable_extended_keys = (int)unibi_add_ext_str(ut, "ext.disable_extended_keys",
                                                                   "\x1b[>4m");
  }
  if (!kitty && (vte_version == 0 || vte_version >= 5400)) {
    data->input.supports_modify_other_keys = true;
  }
```

Reply:
A terminal could still support the kitty keyboard protocol even if this `if` condition is true. For example `foot` and `iTerm2` both support it, so we need to query regardless and we use the `modifyOtherKeys` sequence as a fallback.

I guess we could set the `enable_extended_keys` sequences here as you suggest along with `waiting_for_extkeys_response`, and only emit the sequence in `terminfo_start` if `waiting_for_extkeys_response` is 0.

Reply:
Sorry to keep changing my opinion, but how about using `enum` in the current implementation as follows?

```diff
commit 81c10f57f970d6d76c998796a3e2911e04a1cb58
Author: erw7 <erw7.github@gmail.com>
Date:   Sun Apr 24 15:25:48 2022 +0900

    refactor(tui): change to switch sequence by CSIuType

diff --git a/src/nvim/tui/input.c b/src/nvim/tui/input.c
index 696da7835..642214d67 100644
--- a/src/nvim/tui/input.c
+++ b/src/nvim/tui/input.c
@@ -42,6 +42,7 @@ void tinput_init(TermInput *input, Loop *loop)
   input->paste = 0;
   input->in_fd = STDIN_FILENO;
   input->waiting_for_bg_response = 0;
+  input->csiu_type = kCSIuNone;
   // The main thread is waiting for the UI thread to call CONTINUE, so it can
   // safely access global variables.
   input->ttimeout = (bool)p_ttimeout;
@@ -361,7 +362,7 @@ static void tk_getkeys(TermInput *input, bool force)
 
         if (input->waiting_for_csiu_response > 0) {
           if (initial == '?' && command == 'u') {
-            input->supports_csiu = true;
+            input->csiu_type = kCSIuKitty;
           } else if (initial == '?' && command == 'c') {
             input->waiting_for_csiu_response = 0;
             tui_enable_extkeys(input->tui_data);
diff --git a/src/nvim/tui/input.h b/src/nvim/tui/input.h
index 7ddca0d0c..696344815 100644
--- a/src/nvim/tui/input.h
+++ b/src/nvim/tui/input.h
@@ -8,6 +8,13 @@
 #include "nvim/event/time.h"
 #include "nvim/tui/tui.h"
 
+typedef enum {
+  kCSIuNone,
+  kCSIuKitty,
+  kCSIuXterm,
+} CSIuType;
+
+
 typedef struct term_input {
   int in_fd;
   // Phases: -1=all 0=disabled 1=first-chunk 2=continue 3=last-chunk
@@ -16,7 +23,7 @@ typedef struct term_input {
   bool ttimeout;
   int8_t waiting_for_bg_response;
   int8_t waiting_for_csiu_response;
-  bool supports_csiu;
+  CSIuType csiu_type;
   long ttimeoutlen;
   TermKey *tk;
 #if TERMKEY_VERSION_MAJOR > 0 || TERMKEY_VERSION_MINOR > 18
diff --git a/src/nvim/tui/tui.c b/src/nvim/tui/tui.c
index e6fadd1dc..b404be22f 100644
--- a/src/nvim/tui/tui.c
+++ b/src/nvim/tui/tui.c
@@ -186,11 +186,18 @@ void tui_enable_extkeys(TUIData *data)
   unibi_term *ut = data->ut;
   UI *ui = data->bridge->ui;
 
-  if (input.supports_csiu) {
+  if (input.csiu_type == kCSIuKitty) {
     data->unibi_ext.enable_extended_keys = (int)unibi_add_ext_str(ut, "ext.enable_extended_keys",
                                                                   "\x1b[>1u");
     data->unibi_ext.disable_extended_keys = (int)unibi_add_ext_str(ut, "ext.disable_extended_keys",
                                                                    "\x1b[<1u");
+  } else if (input.csiu_type == kCSIuXterm) {
+    // Fallback to modifyOtherKeys if terminal does not support CSI u
+    data->unibi_ext.enable_extended_keys = (int)unibi_add_ext_str(ut, "ext.enable_extended_keys",
+                                                                  "\x1b[>4;2m");
+    data->unibi_ext.disable_extended_keys = (int)unibi_add_ext_str(ut,
+                                                                   "ext.disable_extended_keys",
+                                                                   "\x1b[>4;0m");
   }
 
   unibi_out_ext(ui, data->unibi_ext.enable_extended_keys);
@@ -2101,12 +2108,7 @@ static void augment_terminfo(TUIData *data, const char *term, long vte_version,
   }
 
   if (!kitty && (vte_version == 0 || vte_version >= 5400)) {
-    // Fallback to modifyOtherKeys if terminal does not support CSI u
-    data->unibi_ext.enable_extended_keys = (int)unibi_add_ext_str(ut, "ext.enable_extended_keys",
-                                                                  "\x1b[>4;2m");
-    data->unibi_ext.disable_extended_keys = (int)unibi_add_ext_str(ut,
-                                                                   "ext.disable_extended_keys",
-                                                                   "\x1b[>4;0m");
+    data->input.csiu_type = kCSIuXterm;
   }
 }
 ```